circuit/src/merkle_tree/native.rs

```diff
@@ -0,0 +1,172 @@
+use std::marker::PhantomData;
+
+use halo2wrong::halo2::arithmetic::FieldExt;
+use num_traits::pow;
+use std::collections::HashMap;
+
+use crate::{params::RoundParams, poseidon::native::Poseidon};
+const WIDTH: usize = 5;
+
+#[derive(Clone, Debug)]
+/// MerkleTree structure
+struct MerkleTree<F: FieldExt, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	/// PhantomData for the params
+	_params: PhantomData<P>,
+	/// Variables with level and value to represent nodes
+	nodes: HashMap<usize, Vec<F>>,
+	/// Height of the tree
+	height: usize,
+}
+
+impl<F: FieldExt, P> MerkleTree<F, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	/// Build a MerkleTree from given leaf nodes
+	fn build_tree(mut leaves: Vec<F>, height: usize) -> Self {
+		assert!(leaves.len() <= pow(2, height));
+		// 0th level is the leaf level and the max level is the root level
+		let mut nodes = HashMap::new();
+		// Assign zero to the leaf values if they are empty
+		for i in leaves.len()..pow(2, height) {
+			leaves.push(F::zero());
+		}
+		nodes.insert(0, leaves);
+
+		for level in 0..height {
+			let mut hash = Vec::new();
+			for i in 0..nodes[&level].len() {
+				if i % 2 == 0 {
+					let hasher: Poseidon<F, WIDTH, P> = Poseidon::new([
+						nodes[&level][i],
+						nodes[&level][i + 1],
+						F::zero(),
+						F::zero(),
+						F::zero(),
+					]);
+					hash.push(hasher.permute()[0]);
+				}
+			}
+			nodes.insert(level + 1, hash);
+		}
+		MerkleTree { _params: PhantomData, nodes, height }
+	}
+
+	/// Find path for the given value to the root
+	fn find_path(&mut self, value: F) -> Path<F, P> {
+		let mut value_index = None;
+		for i in 0..self.nodes[&0].len() {
+			if value == self.nodes[&0][i] {
+				value_index = Some(i);
+				break;
+			}
+		}
```

Comment:
To simplify this search, you can use:
```rust
let value_index = self.nodes[&0].iter().position(|x| x == &value);
```

Reply:
I should add that this approach of finding leaf nodes by values will break in case we have leaves with the same value, but we can leave that concern for the future